File: 44.js
function countSafeZones(n, edges, infected, safe) {
    const g = Array.from({length:n},()=>[]);
    for (let [u,v] of edges) {
        g[u].push(v);
        g[v].push(u);
    }

    const INF = 1e9;
    const time = Array(n).fill(INF);
    const q = [];

    for (let x of infected) {
        time[x] = 0;
        q.push(x);
    }

    for (let i=0;i<q.length;i++) {
        let u = q[i];
        for (let v of g[u]) {
            if (time[v] > time[u] + 1) {
                time[v] = time[u] + 1;
                q.push(v);
            }
        }
    }

    const vis = Array(n).fill(false);
    function dfs(u) {
        vis[u] = true;
        let ok = time[u] === INF;
        for (let v of g[u]) {
            if (safe[v] && !vis[v]) {
                ok = ok && dfs(v);
            }
        }
        return ok;
    }

    let ans = 0;
    for (let i=0;i<n;i++) {
        if (safe[i] && !vis[i] && dfs(i))
            ans++;
    }
    return ans;
}

console.log(
    countSafeZones(
        7,
        [[0,1],[1,2],[2,3],[3,4],[4,5],[5,6]],
        [0],
        [false,true,true,false,true,true,true]
    )
);

File: 44.py
from collections import deque

def count_safe_zones(n, edges, infected, safe):
    g = [[] for _ in range(n)]
    for u, v in edges:
        g[u].append(v)
        g[v].append(u)

    INF = 10**9
    time = [INF]*n
    q = deque()

    for x in infected:
        time[x] = 0
        q.append(x)

    while q:
        u = q.popleft()
        for v in g[u]:
            if time[v] > time[u] + 1:
                time[v] = time[u] + 1
                q.append(v)

    visited = [False]*n

    def dfs(u):
        visited[u] = True
        ok = time[u] == INF
        for v in g[u]:
            if safe[v] and not visited[v]:
                ok &= dfs(v)
        return ok

    count = 0
    for i in range(n):
        if safe[i] and not visited[i]:
            if dfs(i):
                count += 1

    return count

# Driver
n = 7
edges = [(0,1),(1,2),(2,3),(3,4),(4,5),(5,6)]
infected = [0]
safe = [False,True,True,False,True,True,True]

print(count_safe_zones(n, edges, infected, safe))
